common/diagnostics/Metrics.java

```diff
@@ -6,203 +6,761 @@
+            primaryDatabaseHashes.remove(databaseHash);
+        }
     }
 
-    public void setCurrentCount(CurrentCounts.Kind kind, long value) {
-        usage.set(kind, value);
+    public void requestSuccess(@Nullable String databaseName, Metrics.NetworkRequests.Kind kind) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        requests.get(databaseHash).success(kind);
     }
 
-    public void registerError(String errorCode) {
-        userErrors.register(errorCode);
+    public void requestFail(@Nullable String databaseName, Metrics.NetworkRequests.Kind kind) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        requests.get(databaseHash).fail(kind);
     }
 
-    protected String formatPrometheus() {
-        return String.join("\n", system.formatPrometheus(), requests.formatPrometheus(), usage.formatPrometheus(), userErrors.formatPrometheus());
+    public void incrementCurrentCount(@Nullable String databaseName, ConnectionPeakCounts.Kind kind) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        databaseLoad.get(databaseHash).incrementCurrent(kind);
+    }
+
+    public void decrementCurrentCount(@Nullable String databaseName, ConnectionPeakCounts.Kind kind) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        databaseLoad.get(databaseHash).decrementCurrent(kind);
+    }
+
+    public void registerError(@Nullable String databaseName, String errorCode) {
+        String databaseHash = hashAndAddDatabaseIfAbsent(databaseName);
+        userErrors.get(databaseHash).register(errorCode);
     }
 
-    protected JsonObject asJSON() {
-        JsonObject metrics = new JsonObject();
-        metrics.add("system", system.asJSON());
-        metrics.add("requests", requests.asJSON());
-        metrics.add("DB usage", usage.asJSON());
-        metrics.add("user errors", userErrors.asJSON());
+    public void submitDatabaseDiagnostics(Set<DatabaseDiagnostics> databaseDiagnostics) {
+        Set<String> deletedDatabases = new HashSet<>(databaseLoad.keySet());
+
+        for (DatabaseDiagnostics diagnostics : databaseDiagnostics) {
+            String databaseHash = hashAndAddDatabaseIfAbsent(diagnostics.databaseName());
+            databaseLoad.get(databaseHash).setSchemaLoad(diagnostics.schemaLoad());
+            databaseLoad.get(databaseHash).setDataLoad(diagnostics.dataLoad());
+            updatePrimaryDatabases(databaseHash, diagnostics.isPrimaryServer());
+            deletedDatabases.remove(databaseHash);
+        }
+
+        deletedDatabases.forEach(databaseHash -> databaseLoad.get(databaseHash).setDatabaseDeleted());
+    }
+    
+    private static String formatDateTime(LocalDateTime dateTime) {
+        return dateTime.truncatedTo(ChronoUnit.SECONDS).toString();
+    }
+
+    protected JsonObject asJSON(boolean reporting) {
+        boolean isNoUsageDiagnostics = reporting && !base.getReportingEnabled();
```

Comment:
`isNoUsageDiagnostics` is a) negative and b) weird. `addUsageDiagnostics` would do here (`doAddUsageDiagnostics` is also acceptable and arguably more obviously boolean)